### ISSUE TITLE
[clang static analyzer] core.NullDereference false negative after endl

### ISSUE THREAD

**Issue by kuchungmsft** [llvm/llvm-project#104119]
```
#include <iostream>

static void before_endl()
{
	int* before_endl = nullptr;
	*before_endl = 0;

	std::cout << std::endl;
}

static void after_endl()
{
	std::cout << std::endl;

	int* after_endl = nullptr;
	*after_endl = 0;
}

int main()
{
    after_endl();
    before_endl();
}
```

See it: https://godbolt.org/z/TY3z6nYqs

In this case, there is no [core.NullDereference] warning for `*after_endl = 0;`

**Comment by haoNoQ** [comment]
This appears to be specific to GNU libstdc++. We identify both crashes correctly if you pass `-stdlib=libc++`. Let me see if we run out of some budget when we try to model the GNU implementation of `std::cout` or `std::endl` "step by step". (Why are we modeling them this way at all? Aren't we supposed to skip that for standard functions?)

**Comment by haoNoQ** [comment]
> For some reason I can't seem to reproduce it locally. 

Oh, I think it was a bad copy-paste. Finally reproduced.

It's the `NoStoreFuncVisitor` for the `after_endl` pointer:
```
   495 	    // We make an exception for system header functions that have no branches.
   496 	    // Such functions unconditionally fail to initialize the variable.
   497 	    // If they call other functions that have more paths within them,
   498 	    // this suppression would still apply when we visit these inner functions.
   499 	    // One common example of a standard function that doesn't ever initialize
   500 	    // its out parameter is operator placement new; it's up to the follow-up
   501 	    // constructor (if any) to initialize the memory.
   502 	    if (!N->getStackFrame()->getCFG()->isLinear()) {
   503 	      static int i = 0;
-> 504 	      R.markInvalid(&i, nullptr);
   505 	    }
```

I.e., we think that we're exploring an execution path on which one of the functions in the expression `std::cout << std::endl` was _responsible_ for setting `after_endl` to a non-null value, but somehow _failed_ to do that. Because it's a "well-written" standard library function that cannot possibly have bugs of that nature, this obviously cannot happen, so the suppression kicks in.

So I have two questions here.
1. Why do we think the function call `std::cout << std::endl` was _responsible_ for initializing `after_endl`? Did we properly check for that before suppressing? IIUC this check is actually performed *last*, in `maybeEmitNoteForParameters()`. Maybe this suppression should only be applied after we've confirmed that the note would be emitted?
2.  Why is the bug visitor even alive at this point, given that we've already tracked the runtime value `after_endl` back to its ultimate origin (the literal `nullptr` in the code) and literally nothing could affect its future runtime value above that point?

@Szelethus this is your favorite code WDYT?
